src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -0,0 +1,450 @@
+        Its main goal is to ensure that the pileup DID has all the requested
+        rules (and nothing beyond them), according to the pileup object
+        configuration. Pileup documents that are updated as a result of this
+        logic should have their data persisted back in MongoDB.
+
+		1. Read pileup document from MongoDB with filter active=true
+		2. if expectedRSEs is different than currentRSEs, then further data placement is required (it's possible that data removal is required!)
+		3. make a local copy of the currentRSEs value to track rules incomplete but ongoing
+		4. for each rule matching the DID + Rucio account, perform:
+		   - if rule RSE is not in expectedRSEs, then this rule needs to be deleted.
+		     - upon successful rule deletion, also remove the RSE name from currentRSEs and ruleIds (if any)
+		     - make a log record
+		   - else, save this rule RSE in the local copy of currentRSEs. This rule is likely still being processed by Rucio.
+		5. now that we evaluated expected versus current, for each expectedRSEs not in our local copy of currentRSEs:
+		   - first, check whether the RSE has enough available space available for that (we can assume that any storage with less than 1 TB available cannot be considered for pileup data placement)
+		     - in case of no space available, make a log record
+		   - in case there is enough space, make a Rucio rule for that DID + Rucio account + RSE
+		     - now append the rule id to the ruleIds list
+		6. once all the relevant rules have been created - or if there was any changes to the pileup object - persist an up-to-date version of the pileup data structure in MongoDB
+
+        """
+        spec = {'active': True}
+        docs = self.mgr.getPileup(spec)
+        taskSpec = self.getTaskSpec()
+        taskSpec['marginSpace'] = marginSpace
+        asyncio.run(performTasks('active', docs, taskSpec))
+
+    def getTaskSpec(self):
+        """Return task spec"""
+        spec = {'manager': self.mgr, 'logger': self.logger, 'report': self.report,
+                'rucioClient': self.rucioClient, 'rucioAccount': self.rucioAccount}
+        return spec
+
+    def getReport(self):
+        """
+        Return report object to upstream codebase
+        """
+        return self.report
+
+
+def rucioArgs(account):
+    """Return relevant rucio arguments"""
+    kwargs = {'groupping': 'ALL', 'account': account,
+              'lifeTime': None, 'locked': False, 'notify': 'N',
+              'purge_replicas': False, 'ignore_availability': False,
+              'ask_approval': False, 'asynchrounous': True, 'priority': 3}
+    return kwargs
+
+
+def monitoringTask(doc, spec):
+    """
+    Perform single monitoring task over provided MSPileup document
+
+    :param mgr: MSPileupData manager
+    :param doc: MSPileup document
+    :param spec: task spec dict
+    """
+    mgr = spec['manager']
+    uuid = spec['uuid']
```

Comment:
The uuid has nothing to do with stable service or production environment. It is a way to simplify debugging process. When you execute concurrent tasks you need to have uid of the task. It can be anything. In our case it can be pileup name. But pileup name has slashes and when you debug a log dealing with slashes in vim,less,more is tedious process as they need to be escaped. So, the uuid will allow to simplify debugging. By knowing uuid you can "trace" specific task, you may see timestamp of log entries and follow the path of execution. Truest me I debug DAS for many years and found UUID is extremely useful concept for that.

Reply:
we decide it to keep it around.